lisa/src/test/java/it/unive/lisa/test/imp/tutorial/Parity.java

```diff
@@ -0,0 +1,220 @@
+package it.unive.lisa.test.imp.tutorial;
+
+import it.unive.lisa.analysis.SemanticDomain.Satisfiability;
+import it.unive.lisa.analysis.SemanticException;
+import it.unive.lisa.analysis.nonrelational.BaseNonRelationalValueDomain;
+import it.unive.lisa.cfg.type.Type;
+import it.unive.lisa.symbolic.value.BinaryOperator;
+import it.unive.lisa.symbolic.value.Constant;
+import it.unive.lisa.symbolic.value.Identifier;
+import it.unive.lisa.symbolic.value.TernaryOperator;
+import it.unive.lisa.symbolic.value.UnaryOperator;
+
+public class Parity extends BaseNonRelationalValueDomain<Parity> {
+
+	private enum Values {
+		ODD,
+		EVEN,
+		TOP,
+		BOT
+	}
+
+	private final Values parity;
+
+	private Parity(Values parity) {
+		this.parity = parity;
+	}
+
+	public Parity() {
+		this(Values.TOP);
+	}
+
+	@Override
+	public Parity top() {
+		return new Parity(Values.TOP);
+	}
+
+	@Override
+	public boolean isTop() {
+		return getParity() == Values.TOP;
+	}
+
+	@Override
+	public boolean isBottom() {
+		return getParity() == Values.BOT;
+	}
+
+	@Override
+	public Parity bottom() {
+		return new Parity(Values.BOT);
+	}
+
+	@Override
+	public String representation() {
+		switch (parity) {
+		case BOT:
+			return "BOT";
+		case ODD:
+			return "Odd";
+		case EVEN:
+			return "Even";
+		default:
+			return "TOP";
+		}
+	}
+
+	public Values getParity() {
+		return parity;
+	}
+
+	@Override
+	protected Parity evalNullConstant() {
+		return top();
+	}
+
+	@Override
+	protected Parity evalNonNullConstant(Constant constant) {
+		if (constant.getValue() instanceof Integer) {
+			Integer i = (Integer) constant.getValue();
+			return i % 2 == 0 ? even() : odd();
+		}
+
+		return top();
+	}
+
+	private Parity odd() {
```

Comment:
These are all singletons. The analysis scales better if we avoid creating new objects